### PR TITLE
Add structural equality and order comparison for common collections

### LINKED ISSUE #77209: [API Proposal]: Add structural equality and order comparison for common collections
### Background and motivation

This proposal attempts to consolidate a number of issues requesting `IEqualityComparer<T>` and `IComparer<T>` combinators acting on common collection types, e.g. #33873, #44796, https://github.com/dotnet/runtime/issues/77183#issuecomment-1283654919 and #19644. Defining structural equality/order comparison on collection types is a common requirement, for example when working with incremental source generators.

### API Proposal

```csharp
namespace System.Collections.Generic;

public static class EqualityComparer
{
    // Equality comparison using sequence equality
    public static IEqualityComparer<IEnumerable<T>> CreateEnumerableComparer<T>(IEqualityComparer<T> elementComparer = null);
    public static IEqualityComparer<ReadOnlyMemory<T>> CreateMemoryComparer<T>(IEqualityComparer<T> elementComparer = null);
    public static IEqualityComparer<IReadOnlyList<T>> CreateListComparer<T>(IEqualityComparer<T> elementComparer = null);

    // Equality comparison à la HashSetEqualityComparer
    // cf. https://github.com/dotnet/runtime/blob/5ef4c68b08b5f55df91ce62cca86babfdb321162/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/HashSetEqualityComparer.cs
    public static IEqualityComparer<IReadOnlySet<T>> CreateSetComparer<T>(IEqualityComparer<T> elementComparer = null);
    public static IEqualityComparer<IReadOnlyDictionary<TKey, TValue>> CreateDictionaryComparer<TKey, TValue>(IEqualityComparer<TKey> keyComparer= null, IEqualityComparer<TValue> valueComparer = null);
}

public static class Comparer
{
    // Comparison using lexicographic ordering
    public static IComparer<IEnumerable<T>> CreateEnumerableComparer<T>(IComparer<T> elementComparer = null);
    public static IComparer<ReadOnlyMemory<T>> CreateMemoryComparer<T>(IComparer<T> elementComparer = null);
    public static IComparer<IReadOnlyList<T>> CreateListComparer<T>(IComparer<T> elementComparer = null);
}
```

### API Usage

```csharp
IEqualityComparer<string[]> stringArrayComparer = EqualityComparer.CreateEnumerableComparer(StringComparer.InvariantCultureIgnoreCase);
var dictionary = new Dictionary<string[], object>(stringArrayComparer);

dictionary.Add(new string[] { "key" }, null);
dictionary.ContainsKey(new string[] { "KEY" }); // true
```

```csharp
IComparer<string[]> stringArrayComparer = Comparer.CreateEnumerableComparer(StringComparer.InvariantCultureIgnoreCase);
var dictionary = new SortedDictionary<string[], object>(stringArrayComparer);

dictionary.Add(new string[] { "key" }, null);
dictionary.ContainsKey(new string[] { "KEY" }); // true
```

### Alternative Designs

Making the combinators generic on the collection type would allow construction-time specialization and help with [devirtualization](https://sharplab.io/#v2:C4LghgzgtgPgAgBgARwIwDoDCB7ANrgUwGNgBLbAOwgG4BYAKERVQBY7Hk10AlAVwrJQCWbFAAOpQgCcAygSkA3UkQI0GTGQAswUsQBkwAIx79BBdgHoLTLjnzEylCOgCSUKL2BHC1JFaQU2MBIYApgkt4ESKQUSBDaurhGDOoAzCgATEiYSAwA3gxIRSjpaABs0QJIAOIEwHoEFADmwJoAFG4eXoaEAIJSUmAAngA8McAAfEi4pBDAAJSFxQX0xWsoAOzTs8AipuxrAL5L66dn52snSADaAFKkwLUU8sptwENiBNgAZh3unpF+oNRuMJvN5gBdK5wUqoCrjGp1BrNVojAAqEw6eh22zmi1WyyuazgWxmcz2AgOxWOBKKVzuDyeLyIbw+X1+nQBPQIQOGYwEYIANEh3p8fm1xuCobSSsx4VVavVGi1NOjsXNhRi2mj1cEyQsicUAO6aeRRHU4kBIFzcAhgAAmAHkKLghrr0RMritziTcbscPsrjToekpHb7ZRXXFgFJeCRrf9un0BnyMUgrTbw87Xe6MV7DUUYUg0SLNLNrgiYvaCAAPCFIAC8U1aUmwRoCBHbXAAckE3GJCEIBAR7QBRGsqMSOChteZU04F2UIgNVJul1vt55djC94D9weNYAj8eT6ez9iLosuUcUXhCQbAbBSD2I4A3u/yMCPqSzxfei5FC2bYdtu6C7vuBBDkeY4TgQU7kDOc6LsGMprNet73l+T7MKk6CKu+mHfr+qGEiRpxAZunbMGBfbiAew4waeCHnshQYMIcQA==), e.g.

```C#
public static IEqualityComparer<TEnumerable> CreateEnumerableComparer<TEnumerable, TElement>(IEqualityComparer<TElement> elementComparer = null)
     where TEnumerable : IEnumerable<TElement>
{
     elementComparer ??= EqualityComparer<T>.Default;

     if (typeof(IReadOnlyList<TElement>).IsAssignableFrom(typeof(TEnumerable)))
     {
          return new ListSequenceComparer<TEnumerable, TElement>(elementComparer));
     }

     return new EnumerableSequenceComparer<TEnumerable, TElement>(elementComparer);
}
```

But obviously that would make calling the combinators more complicated:

```C#
IEqualityComparer<int[]> arrayComparer = EqualityComparer.CreateEnumerableComparer<int[], int>();
```
Full alternative proposal:

```csharp
namespace System.Collections.Generic;

public static class EqualityComparer
{
    // Equality comparison using sequence equality
    public static IEqualityComparer<TEnumerable> CreateEnumerableComparer<TEnumerable, T>(IEqualityComparer<T> elementComparer = null) 
        where TEnumerable : IEnumerable<T> { }

    public static IEqualityComparer<ReadOnlyMemory<T>> CreateMemoryComparer<T>(IEqualityComparer<T> elementComparer = null) { }

    // Equality comparison à la HashSetEqualityComparer
    // cf. https://github.com/dotnet/runtime/blob/5ef4c68b08b5f55df91ce62cca86babfdb321162/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/HashSetEqualityComparer.cs
    public static IEqualityComparer<TSet> CreateSetComparer<TSet, T>(IEqualityComparer<T> elementComparer = null) 
        where TSet : IReadOnly<T> { }

    public static IEqualityComparer<TDictionary> CreateDictionaryComparer<TDictionary, TKey, TValue>(IEqualityComparer<TKey> keyComparer= null, IEqualityComparer<TValue> valueComparer = null) 
        where TDictionary : IDictionary<TKey, TValue> { }
}

public static class Comparer
{
    // Comparison using lexicographic ordering
    public static IComparer<TEnumerable> CreateEnumerableComparer<TEnumerable, T>(IComparer<T> elementComparer = null) 
        where TEnumerable : IEnumerable<T> { }

    public static IComparer<ReadOnlyMemory<T>> CreateMemoryComparer<T>(IComparer<T> elementComparer = null) { }
}
```

### Risks

_No response_

### PR DESCRIPTION
**Pull request by manandre**
Closes #77209

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/Collections/Generic/EqualityComparer.cs [hunk 1]
```diff
@@ -262,4 +260,259 @@ public void GetObjectData(SerializationInfo info, StreamingContext context)
...
+            new SetEqualityComparer<TSet, T>(elementComparer);
+
+        public static IEqualityComparer<TDictionary> CreateDictionaryComparer<TDictionary, TKey, TValue>(IEqualityComparer<TKey>? keyComparer = null, IEqualityComparer<TValue>? valueComparer = null)
+            where TDictionary : IReadOnlyDictionary<TKey, TValue> =>
+            new DictionaryEqualityComparer<TDictionary, TKey, TValue>(keyComparer, valueComparer);
+    }
+
+    [Serializable]
+    [TypeForwardedFrom("mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089")]
+    // Needs to be public to support binary serialization compatibility
```
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/Collections/Generic/EqualityComparer.cs:283]: This is a new type. There are no binary compatibility concerns with new types. We explicitly do not want to make any new types compatible with binary serialization.

The type should be internal and all `TypeForwardedFrom` and `Serializable` attributes should be deleted from this change.

Comment (manandre) [reply src/libraries/System.Private.CoreLib/src/System/Collections/Generic/EqualityComparer.cs]: Ok, I was not sure.
Removed!

File: src/libraries/System.Private.CoreLib/src/System/Collections/Generic/EqualityComparer.cs [hunk 2]
```diff
@@ -262,4 +260,278 @@ public void GetObjectData(SerializationInfo info, StreamingContext context)
...
+            //     y is Dictionary<TKey, TValue> dicty && dicty.Comparer.Equals(_keyComparer))
+            // {
+            //     foreach ((TKey yKey, TValue yValue) in y)
+            //     {
+            //         if (!x.TryGetValue(yKey, out TValue xValue) || !_valueComparer.Equals(yValue, xValue))
+            //             return false;
+            //     }
+            //
+            //     return true;
+            // }
```
Comment (manandre) [inline src/libraries/System.Private.CoreLib/src/System/Collections/Generic/EqualityComparer.cs:494]: This optimization cannot be activated here, as the `Dictionary<TKey, TValue>` class has the constraint `where TKey : notnull` which cannot be enforced here.
Is there a way to solve this?

File: src/libraries/System.Collections/ref/System.Collections.cs [hunk 3]
```diff
@@ -556,6 +556,25 @@ public abstract partial class EqualityComparer<T> : System.Collections.Generic.I
...
+            where TEnumerable : System.Collections.Generic.IEnumerable<T> { throw null; }
+
+        public static System.Collections.Generic.IEqualityComparer<TSet> CreateSetComparer<TSet, T>(System.Collections.Generic.IEqualityComparer<T>? elementComparer = null) 
+            where TSet : System.Collections.Generic.IReadOnlySet<T> { throw null; }
+
+        public static System.Collections.Generic.IEqualityComparer<TDictionary> CreateDictionaryComparer<TDictionary, TKey, TValue>(System.Collections.Generic.IEqualityComparer<TKey>? keyComparer = null, System.Collections.Generic.IEqualityComparer<TValue>? valueComparer = null) 
+            where TDictionary : System.Collections.Generic.IReadOnlyDictionary<TKey, TValue> { throw null; }
+    }
+
+    public static class ComparerFactory
```
Comment (manandre) [inline src/libraries/System.Collections/ref/System.Collections.cs:572]: The approved API uses the `Comparer` class in the `System.Collections.Generic` namespace, but it is [already present](https://learn.microsoft.com/en-us/dotnet/api/system.collections.comparer) in the `System.Collections` namespace
https://github.com/dotnet/runtime/blob/5b8ebeabb32f7f4118d0cc8b8db28705b62469ee/src/libraries/System.Runtime/ref/System.Runtime.cs#L7251-L7258

Issue reported [here](https://github.com/dotnet/runtime/issues/77209#issuecomment-1356557303)

**Comment by manandre** [comment]
PR converted to draft as the approved API needs rework.

**Pull request closed without merging**